Method stLearn

peicai opened the pull request:
Added a script for stLearn implementation, a yml and 13 config files;  linked to issue https://github.com/SpatialHackathon/SpaceHack2023/issues/6.

Note: 11 out of 13 config files require the inclusion of image paths, specifically for 'H_E.tiff' and 'H_E.json'. The proper formatting example for these image paths is as follows: --image $data_dir/H_E.tiff $data_dir/H_E.json. In Snakemake workflows, the function `get_sample_image` in the `methods.smk` file might need to be modified.

peicai committed:
Add new method - stLearn

Not tested yet

peicai committed:
Add new method - stLearn

Not tested yet

peicai committed:
fix a typo bug in dir

peicai committed:
fix a typo bug in dir

peicai committed:
Add 2 methods: STAGATE and GraphST

peicai committed:
Add 2 methods: STAGATE and GraphST

peicai committed:
Merge branch 'main' into method_stLearn_pcai

peicai committed:
stLearn implemented

peicai committed:
stLearn implemented

peicai committed:
resolve conflicts

niklasmueboe commented on method/stLearn/method_stLearn.py:
```diff
@@ -0,0 +1,235 @@
+else:
+    # Calculate and use PCA
+    st.pp.scale(adata)
+    st.em.run_pca(adata, n_comps=config["n_comps"], random_state=seed)
```
Is this just a norml PCA?

If yes we don't need to handle this case and assume that you are always getting a reduced dimensionality passed via args

peicai replied:
Yes, it is just a normal PCA. Ok, I'll delete these.

niklasmueboe commented on method/stLearn/method_stLearn.py:
```diff
@@ -0,0 +1,235 @@
+                    spot_diameter_fullres = scale_factors.get('spot_dimeter_fullres', 50.0)
+                else:
+                    scale = 1.0
+                    spot_diameter_fullres = 50.0
```
I am wondering whether we should rather fail than trying to guess parameters?

peicai replied:
Yes, but in the H_E.json file for LIBD_DLPFC data, it only provides the scale parameter...
Should we handle failures and return a message requesting such parameters?

niklasmueboe commented on method/stLearn/method_stLearn.py:
```diff
@@ -0,0 +1,235 @@
+    st.tl.clustering.kmeans(adata,n_clusters=n_clusters, use_data=use_data, key_added="cluster", random_state=seed)
+elif config["method"] == "louvain":
+    st.pp.neighbors(adata,n_neighbors=config["n_neighbors"],use_rep=use_data, random_state=seed)
+    st.tl.clustering.louvain(adata, resolution=config["resolution"], key_added="cluster", random_state=seed)
```
I think for other methods, we handled louvain (/leiden) by scanning the resolution to find n_clusters rather than using a config provided resolution

shdam replied:
> I think for other methods, we handled louvain (/leiden) by scanning the resolution to find n_clusters rather than using a config provided resolution

Was also curious about this. Found the implementation [here](https://github.com/SpatialHackathon/SpaceHack2023/blob/ac2784abacf5342b13317be7e1b7cafdb716a4cb/method/SEDR/SEDR_method.py#L123)

For Omnibenchmark, I implemented this similar R version:
`utils_tunek <- function(snn, k){
  
  for(r in seq(0.1, 5, by = 0.05)) {
    res <- igraph::cluster_leiden(graph = snn, resolution_parameter = r)
    if(res$nb_clusters >= k) break
  }
  return(res)
}
`

Jieran-S replied:
I was a bit suspicious about this implementation using grid search, what if the number of clusters is outside the search range? --maybe we can keep in mind this could be the source of future errors.

niklasmueboe reviewed (changes requested):
It looks like some unrelated files got deleted. Please restore.

peicai committed:
rebase branch on 'ac2784a'

peicai committed:
remove PCA computation and fixed image params; tune louvain resolution from config

peicai committed:
tune louvain resolution from configs

peicai committed:
resolve conflicts: add BayesSpace

peicai commented:
I think that's because there are always some conflicts. I'll try to keep only stLearn and see what happens.

Commit:
update all files outside stLearn

peicai commented:
Regarding SpaGCN configs, I might have deleted them to resolve conflicts somehow. I've updated all files in the method from the main branch, hoping that it will make merging easier.

> Can you remove tracking on all files outside stLearn
I removed tracking on all other files, but I'm not sure if I did it correctly or if it will work.

shdam commented:
Oh no, that made it much worse - you need to remove tracking of all files outside stLearn, not add tracking.

Something like 
`git reset --soft <commit>` to a specific commit without external changes and then possibly remove staged files outside stLearn with:
`git restore --staged <files>`

shdam commented:
The goal is to reduce [Files changed](https://github.com/SpatialHackathon/SpaceHack2023/pull/179/files) to only have stLearn files. Otherwise, merging will be a mess.
Hope it makes sense :)

Commit:
removed tracking of all files outside stLearn

Jieran-S commented:
Hi @peicai, I would close this PR and when u have time could you branch the native branch and update the implementation of stLearn in the format of native implementation (I think you basically get it done) and submit a PR to the native branch? Thanks!
